[HM] Final MPL Link

FZill opened the pull request:
Linked fluid_density, permeabiltiy and ref. temperature in HM to MPL.
Fluid compressibiltiy is now properly calculated as drho_dp / rho.
Also filled the argument variables with temperature and pressure, so that MPL can calculate e.g. the IdealGasLaw.
Finally the FluidType class, which was only temporarily used by HM, could be removed.
HM-models which used the ideal gas law now have to state the molar_mass. To get the same results as before it can be calculated via: M = rho_fr * R * T / p with the fluid_density rho_fr, the specific_gas_constant R, the reference_temperature T and the initial_pressure p.

I updated the benchmarks with this script: 
[ConvertToMPL.txt](https://github.com/ufz/ogs/files/4159077/ConvertToMPL.txt)
It should ged rid of the now unnecessary parameters and add the needed properties.

endJunction commented:
I suggest to move the ideal gas laws in own PR. There are some unit tests and documentation associated. And is a stand alone feature.

TomFischer commented on MaterialLib/MPL/Properties/CompressibilityIdealGasLaw.cpp:
```diff
@@ -0,0 +1,81 @@
+        return 0;
+    }
+
+    OGS_FATAL(
```
What about mixed second variables derivatives (d2p/(dp dT) or d2p/(dT dp))? Should these combination result in an OGS_FATAL?

TomFischer commented on MaterialLib/MPL/Properties/CompressibilityIdealGasLaw.h:
```diff
@@ -0,0 +1,72 @@
+ *              http://www.opengeosys.org/project/license
+ *
+ *  \file
+ *  Created on November 29, 2017, 2:03 PM
```
Probably you can remove this line?

wenqing commented on MaterialLib/MPL/Properties/CompressibilityIdealGasLaw.cpp:
```diff
@@ -0,0 +1,81 @@
+    const double pressure = std::get<double>(
+        variable_array[static_cast<int>(Variable::phase_pressure)]);
+
+    return (1 / pressure);
```
or
``` 
      // return 1/p  
     return 1.0 / std::get<double>(
        variable_array[static_cast<int>(Variable::phase_pressure)]);
```

FZill commented:
Since the compressibiltiy of the fluid can be calculated as drho_dp / rho, there was no need for the second IdealGasLaw-stuff. So this PR got a bit smaller and I made the commits more precise.

endJunction commented on ProcessLib/HydroMechanics/CreateHydroMechanicsProcess.cpp:
```diff
@@ -176,40 +162,11 @@ std::unique_ptr<Process> createHydroMechanicsProcess(
-        fluid_compressibility, reference_temperature,
-        specific_gas_constant, fluid_type};
+        initial_stress,        specific_body_force,
+        reference_temperature};
```
The reference temperature is also available in MPL... https://github.com/ufz/ogs/blob/master/MaterialLib/MPL/PropertyType.h#L64

FZill replied:
also add this a medium property?

endJunction replied:
Yes.

endJunction commented on ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h:
```diff
@@ -215,28 +216,34 @@ void HydroMechanicsLocalAssembler<ShapeFunctionDisplacement,
+                .template dValue<double>(vars, MPL::Variable::phase_pressure,
+                                         x_position, t) / rho_fr;
+
+        auto const S = gas.hasProperty(MPL::PropertyType::storage)
```
let's call it `storage`.

FZill replied:
okay and I noticed it has to be a medium property

FZill commented on ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h:
```diff
@@ -216,6 +214,13 @@ void HydroMechanicsLocalAssembler<ShapeFunctionDisplacement,
         auto& eps = _ip_data[ip].eps;
         auto const& sigma_eff = _ip_data[ip].sigma_eff;
 
+        auto const T_ref =
```
Has somebody a better idea?
I could make T_ref a required property, but since it is not needed in most of the cases I don't think that makes sense.

FZill commented on ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h:
```diff
@@ -216,6 +214,13 @@ void HydroMechanicsLocalAssembler<ShapeFunctionDisplacement,
+                ? medium->property(MPL::PropertyType::reference_temperature)
+                      .template value<double>(vars, x_position, t)
+                : std::numeric_limits<double>::quiet_NaN();
+
```
Does somebody have a better idea?
I could make T_ref a required property, but since it is not needed in most of the cases I don't think that makes sense

TomFischer replied:
Is T_ref independent from the integration point? Then it could be evaluated only once outside of the integration loop.
If T_ref is additionally independent from time it could be evaluated in the constructor and stored in the local assembler. Or if it is even independent from the mesh element? Then, is it possible to move it out of the class and pass it as an argument to the local assembler?

FZill replied:
Since the user could in theory make t_ref spatially dependent (e.g. cavern vertical temperature gradient) it is dependent of the IP, right?
I think with the time dependence it's the same. 
I could move it out of the loop when we assume it is constant per element.
But since it is probably a rare scenario having something else than a constant T_ref in a HM-Process, we maybe can move it further out. But then we could just leave it as a variable in process_data...

FZill replied:
Consensus is, to make T_ref a required property

Scinopode commented on ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h:
```diff
@@ -215,28 +214,41 @@ void HydroMechanicsLocalAssembler<ShapeFunctionDisplacement,
+                .template dValue<double>(vars, MPL::Variable::phase_pressure,
+                                         x_position, t) / rho_fr;
+
+        auto const storage = medium->hasProperty(MPL::PropertyType::storage)
```
This conditional means that you can either use a constant _storage_ parameter (combined fluid and solid compressibilities along with some poro-mechanical magic and porosity) OR to use the concept of compressibilities in the first place. This is OK, but I would highly recommend to get rid of the _storage_ parameter at all! Actually, there is no cost in doing so, since all of the parameters _storage_ is made of have to be provided anyway: fluid compressibility is given via the derivative of the density parameter, porosity and biot-coefficient are required, and solid phase compressibility can be taken from Youngs modulus (or defined directly). I see no benefit using a composite _storage_ parameter; otherwise, using _storage_ limits you in your choice of your other properties (you would have to reevaluate the value of _storage_ when changing porosity, biot-coeffient, or even the ambient temperature).
Please correct me if I'm wrong or if I missed something.

FZill replied:
No you're right. we have just one test which used the specific storage until I replaced it with the corresponding compressibility but there was somebody which wanted the possibility back to give the storage as a user input and I thought I could now do exactly this. But you're right, it is not beneficial. I will revert this.

FZill replied:
storage term is now calculated as before.
And since the compressibility is now fixed as drho_dp the benchmark InjectionProduction1D was inccompatible with the current implementation. I could recreate the same storage with the correct rho_fr = f(p) but since rho_fr is then not constant it influences the results. Thus I updated the solution files of this test. Since it mainly tests the equality of the monolithic and staggered theme solutions this shouldn't matter much.

endJunction commented on Tests/Data/HydroMechanics/Linear/Confined_Compression/cube_1e3.prj:
```diff
@@ -59,6 +61,18 @@
+                <property>
+                    <name>reference_temperature</name>
+                    <type>Constant</type>
+                    <value>0</value>
```
This is rather cold environment...
Should be mentioned at least in the PR. Better would be something around 20 degree Celsius.

FZill replied:
Okay, then I go with the 293.15 K for all tests

endJunction commented on Tests/Data/HydroMechanics/StaggeredScheme/InjectionProduction1D/InjectionProduction1D.prj:
```diff
@@ -71,6 +76,18 @@
+                <property>
+                    <name>reference_temperature</name>
+                    <type>Constant</type>
+                    <value>295.15</value>
```
It's not 293.15K ?

FZill replied:
The reference_temp. is in K, so that means 20°C what's wring with this?

endJunction commented on Tests/Data/HydroMechanics/Verification/hm1_3Dgravity.prj:
```diff
@@ -275,71 +340,13 @@
-        </parameter>
-        <parameter>
-            <name>k</name>
-            <type>Group</type>
```
I'd keep the Group type parameter. There is no need to repeat all other properties four times, just because one parameter is heterogeneous.

FZill replied:
But either way I have to mention in every medium either the constant value or the reference to the parameter, so the medium section will be large anyway. Then I thought, it would be best to choose the constant value, to get rid of the large parameter section.

endJunction commented on Tests/Data/HydroMechanics/StaggeredScheme/InjectionProduction1D/InjectionProduction1D.prj:
```diff
@@ -48,6 +43,16 @@
+                                <independent_variable>
+                                    <variable_name>phase_pressure</variable_name>
+                                    <reference_condition>0</reference_condition>
+                                    <slope>3.3332941333333e-5</slope>
```
Why is the slope different from the fluid's compressibility?

FZill replied:
The compressebility of 3.333... before probably was misscalculated from my side but did not seem to influence the result enough to fail. This value now should correctly result in a storage of 1e-5.

FZill commented:
I updated the benchmarks to have a reasonable value for the reference temperature,
made use of #2786 to shrink down the definition of the 3Dgravity benchmark
and updated the checkRequiredProperties to do check on each element and also added another check to verify if a medium corresponding to the material_ID of the current element exists.
Without the latter, ogs can abort at this position when no media are defined for all materialID's in the mesh.
